compaction/compaction.cc

```diff
@@ -609,7 +610,8 @@ class compaction {
         sstable_writer_config cfg = _table_s.configure_writer("garbage_collection");
         cfg.run_identifier = gc_run;
         cfg.monitor = monitor.get();
-        auto writer = sst->get_writer(*schema(), partitions_per_sstable(), cfg, get_encoding_stats());
+        uint64_t estimated_partitions = std::max(1UL, uint64_t(ceil(partitions_per_sstable() * _estimated_droppable_tombstone_ratio)));
+        auto writer = sst->get_writer(*schema(), estimated_partitions, cfg, get_encoding_stats());
```

Comment:
@raphaelsc don't we need to do the same (just mirrored) to the normal writer?

Reply:
let's say that you have 10 input sstables, each avg'ing 1k partitions, and droppable ratio is 0.5

so we sum input partition count, and we get ~10k. we divide by 10, and we have the estimated partition per sstable of 1k.

if we apply droppable ratio (1 - 0.5 = 0.5), we'll say that each output sstable (there will be ~5 due to 0.5 droppable ratio) will be able to hold 0.5k partitions, which is incorrect.

but I think it's ok to apply (1 - droppable_ratio) if estimated_sstables(output) == 1.

makes sense?

Reply:
for GC-generated sstables, it's ok to apply droppable_ratio because we'll cut the GC-sstable on each exhausted input sstable. 

```
assume droppable_ratio = 0.1

disjoint input:
[ sst 1 ] [ sst 2 ] [ sst 3 ] 

then as we finish compacting [sst 1] 

we'll have something like this: 

         [ sst 2 ] [ sst 3 ] [ gc sstable 1 = 10% of sst 1 ]; note 90% is not enough to fill an entire sstable
         
then 

                    [ sst 3 ] [ gc sstable 1 = 10% of sst 1 ] [ gc sstable 2 = 10% of sst 2 ] [ sst 4 is produced as max_size = 1G is reached ]
                    
and so on
```